lib/launcher.rb

```diff
@@ -147,13 +183,91 @@ def write_pid
       retry
     end
 
-    def webrick_server(app, addresses, port)
+    def add_puma_server_callback(sd_notify)
+      events = ::Puma::Events.new(::Proxy::LogBuffer::Decorator.instance, ::Proxy::LogBuffer::Decorator.instance)
+      events.register(:state) do |status|
+        if status == :running
+          sd_notify.ready_all { sd_notify.status_if_active("Finished launching #{sd_notify.total} instances, ready to serve requests", logger) }
+          sd_notify.status_if_active("Finished launching an instance, #{sd_notify.pending} instances to go...", logger) if sd_notify.pending > 0
+        end
+      end
+      events
+    end
+
+    def format_ip_for_url(address)
+      addr = IPAddr.new(address)
+      addr.ipv6? ? "[#{addr}]" : addr.to_s
+    rescue IPAddr::InvalidAddressError
+      address
+    end
+
+    def add_puma_server(app, address, port, conn_type, sd_notify)
+      address = format_ip_for_url(address)
+      logger.debug "Launching Puma listener at #{address} port #{port}"
+      if conn_type == :ssl
+        require 'cgi'
+        query_list = app[:SSLArgs].to_a.map do |x|
+          "#{CGI::escape(x[0].to_s)}=#{CGI::escape(x[1])}"
+        end
```

Comment:
We should make ForemanRequestFactory.query_string a static method and reuse it:
https://github.com/theforeman/smart-proxy/blob/ebdde78c329d6bf67b198446dae64d8952d0ac4a/lib/proxy/request.rb#L13-L15

At least we can use it as a map
```suggestion
        query_list = app[:SSLArgs].map do |k, v|
          "#{CGI::escape(k.to_s)}=#{CGI::escape(v)}"
        end
```

This results into an array:
```
[1] pry(main)> {a: 1, b: 2}.map { |k, v| "#{k} => #{v}" }
=> ["a => 1", "b => 2"]
```

Reply:
Ahhhh, you are not making it easier. Moved this to `util.rb` helper.